syntax = "proto3";

package chat;

service ChatService {
  rpc CreateAccount(CreateAccountRequest) returns (Response) {}
  rpc Login(LoginRequest) returns (LoginResponse) {}
  rpc ListAccounts(ListAccountsRequest) returns (ListAccountsResponse) {}
  rpc SendMessage(SendMessageRequest) returns (Response) {}
  rpc ReadMessages(ReadMessagesRequest) returns (ReadMessagesResponse) {}
  rpc DeleteMessages(DeleteMessagesRequest) returns (Response) {}
  rpc DeleteAccount(DeleteAccountRequest) returns (Response) {}
  rpc Heartbeat(HeartbeatRequest) returns (Response) {}
}

service ReplicationService {
  rpc ReplicateOperation(ReplicationRequest) returns (Response) {}
  rpc AddReplica(AddReplicaRequest) returns (Response) {}
  rpc TransferState(Empty) returns (StateSnapshot) {}
}

message CreateAccountRequest {
  string username = 1;
  string password = 2;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message ListAccountsRequest {
  string pattern = 1;
}

message SendMessageRequest {
  string sender = 1;
  string recipient = 2;
  string message_text = 3;
}

message ReadMessagesRequest {
  string username = 1;
  int32 num = 2;
  int32 last_seen_id = 3;
}

message DeleteMessagesRequest {
  string username = 1;
  repeated string message_ids = 2;
}

message DeleteAccountRequest {
  string username = 1;
  string password = 2;
}

message HeartbeatRequest {
  string username = 1;
}

message Response {
  string status = 1;
  string message = 2;
}

message LoginResponse {
  string status = 1;
  string message = 2;
  int32 unread_count = 3;
  int32 last_seen_id = 4;
}

message ListAccountsResponse {
  string status = 1;
  string message = 2;
  repeated string accounts = 3;
}

message ReadMessagesResponse {
  string status = 1;
  string message = 2;
  repeated string messages = 3;
}

message ReplicationRequest {
  string op = 1;
  string username = 2;
  string salt = 3;
  string hashed_password = 4;
  string sender = 5;
  string recipient = 6;
  string message_text = 7;
  repeated string message_ids = 8;
  bool forwarded = 9;
}

message AddReplicaRequest {
  string replica_address = 1;
}

message Empty {}

message AccountData {
  string username = 1;
  string salt = 2;
  string hashed_password = 3;
}

message MessageData {
  int32 id = 1;
  string recipient = 2;
  string sender = 3;
  string message_text = 4;
  int32 delivered = 5;
}

message StateSnapshot {
  repeated AccountData accounts = 1;
  repeated MessageData messages = 2;
  int32 max_message_id = 3;
}
